Update openwhisk-knative pipeline to v1beta1

mrutkows opened the pull request:
# Changes

This PR is a complete update of the OpenWhisk to Knative pipeline from v1alpha1 level to v1beta1 resource levels.  This pipeline also adds conditional logic to support additional "branches" of the pipeline to sequence tasks that can detect and build serverless applications for not only NodeJS (previously the only lang. supported), but also Python and Java. In addition, this pipeline utilizes a workspace for sharing data/state across tasks.  Each language branch has its own sample functions with instructions for users to build and test on knative the services they build.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Yaml file complies with [yamllint](https://github.com/adrienverge/yamllint) rules.

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

mrutkows commented:
@pritidesai please can you give your review?

pritidesai commented:
/ok-to-test

mrutkows commented:
This (PR) pipeline and its tasks have no integration tests, but now seem to be blocked by some repo.-wide integration tests?

sbwsg commented:
@mrutkows you'll need to rebase on top of the latest master to get the fixed integration tests.

mrutkows commented:
@sbwsg Hi Scott, even after rebase getting even more (but different assertion type) failures...  not sure how to proceed at this point.

chmouel commented on openwhisk/.gitignore:
```diff
@@ -0,0 +1,46 @@
+##############################################
```
Any chance you can remove this file ?  See here for the reasoning why gitignore should be kept local https://julien.danjou.info/properly-managing-your-gitignore/

mrutkows replied:
@chmouel the design of gitignore supports and encourages use of directory-specific use (see https://labs.consol.de/development/git/2017/02/22/gitignore.html) top allow finer-grained exclusion.  It seems that since the catalog is effectively a collection of subprojects each would want and be encouraged to have its own rules specific to its installation/configuration/tests.  In fact, if you follow the readme's optional instructions to manually runt the examples the scripts and steps cause many new files to be created which we do not want checked int by developers who may be updating this specific catalog entry.  This gitignore reflects those items my fellow developers actually derived from those "oops" moments while working on this specific entry over the last year.

chmouel replied:
I am not so much convinced by this argument of having gitignore everywhere in the catalogs, if the gitignore contained some specific files for the specifics openwhisk files then that would be fine, but those are mostly editors/ide exclude, which are best kept in the user configuration

mrutkows replied:
@chmouel the design of gitignore supports and encourages use of directory-specific use (see https://labs.consol.de/development/git/2017/02/22/gitignore.html) to allow finer-grained exclusion. It seems that since the catalog is effectively a collection of subprojects each would want and be encouraged to have its own rules specific to its installation/configuration/tests.

In fact, if you follow the "openwhisk" specific readme's optional instructions to manually run the examples the scripts and steps cause many new files to be created which we do not want checked int by developers who may be updating this specific catalog entry. This gitignore reflects those items my fellow developers actually derived from those "oops" moments while working on this specific entry over the last year.

However, I agree with the point from the ref. article and removed the "editor" specific entries in the file.  Hopefully, what remains is accurate to the examples only.

mrutkows replied:
> I am not so much convinced by this argument of having gitignore everywhere in the catalogs, if @chmouel As indicated on the update, gitignore is specific to the openwhisk entry and artifacts that are created from following its README only.

mrutkows replied:
@chmouel I decided to co-locate the JSON parameter (input) files in the same subdirectory with the Java pipelineRun YAML as this is the only pipelinerun (example) that uses these input files.

chmouel commented on openwhisk/01-dice-color.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Those .json files are examples, isnt it ? By convention we movet them to the examples/ directory

mrutkows replied:
@chmouel yes, had thought to move them, but it took so much time to get the instructions working that I had hoped to defer this to a subsequent PR as it seemed the priority was to get the resources themselves updated to v1beta1 level.   If this is a "blocker" I can try to find some time in the next several days to do so, but would it suffice to open an issue and assign it to myself against just this catalog entry?

mrutkows replied:
@chmouel I decided to co-locate the JSON parameter (input) files in the same subdirectory with the Java pipelineRun YAML as this is the only pipelinerun (example) that uses these input files.

chmouel commented on openwhisk/tests/taskrun-install-npm-packages.yaml:
```diff
@@ -0,0 +1,55 @@
+apiVersion: tekton.dev/v1alpha1
+kind: PipelineResource
+metadata:
+  name: app-git
```
You only need one PipelineResource to reference, not create two as you have already one in 
https://github.com/tektoncd/catalog/pull/375/files?file-filters%5B%5D=.jpg&file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.png&file-filters%5B%5D=.sh&file-filters%5B%5D=.tmpl&file-filters%5B%5D=.yaml&file-filters%5B%5D=.zip#diff-1c0712a9233a5d7535a1712ebef1ea1aR1

mrutkows replied:
Originally, these "tests" were not meant for automation, but we happened to choose the same subdir. structure.  Removing them from the commit

pritidesai commented on openwhisk/README.md:
````diff
@@ -1,262 +1,569 @@
+    ```bash
+    $ kubectl get pods --namespace kube-system
+    ```
+    <details>
````
Great job @mrutkows hiding details 👍 

![Screen Shot 2020-06-25 at 1 53 50 PM](https://user-images.githubusercontent.com/206285/85794382-62d64200-b6eb-11ea-98cd-8273a9ee73a5.png)

pritidesai commented on openwhisk/README.md:
```diff
@@ -1,262 +1,569 @@
+#### Running the NodeJS example
 
-## Clone this repository
+The `PipelineRun` resource used to build a Python application is named `build-javascript-app-image` and derived from the template file [pipelinerun-build-padding-app.yaml.tmpl](pipelinerun/javascript/pipelinerun-javascript.yaml.tmpl).
```
s/python/javascript/

pritidesai replied:
thanks for fixing this 👍

pritidesai commented on openwhisk/README.md:
````diff
@@ -1,262 +1,569 @@
 
-## Create Service Account for our Runtime Builds
+    ```bash
+    tkn pr describe build-javascript-app-image
````
please replace this with `tkn pr describe build-python-app-image`

pritidesai commented on openwhisk/README.md:
```diff
@@ -1,262 +1,569 @@
+# OpenWhisk Application pipeline for Knative
 
-This directory contains Tekton Task which can be used to Build and Serve Knative compatible applications (i.e., OpenWhisk Actions) on Kubernetes.
+*This catalog offering provides a single pipeline that can be used to build either [Apache OpenWhisk](https://openwhisk.apache.org/) or [Knative](https://openwhisk.apache.org/) compatible containers for supported runtimes used to execute OpenWhisk serverless functions.*
```
please update the knative link here

mrutkows replied:
Thanks @pritidesai.  Silly copy/paste mistake.  Will take the opportunity to add links a couple of other places in the opening paragraphs that may be helpful for easily accessing context.

pritidesai commented on openwhisk/README.md:
```diff
@@ -1,262 +1,569 @@
-- [x] Knative v0.5.0
-- [x] Tekton Pipeline
-- [x] Knative Client (OPTIONAL)
+The project provides a set of [supported language runtimes](https://openwhisk.apache.org/downloads.html#component-releases) that includes a proxy that enforces a *documented contract for function initialization (function injection) and execution* along with a standard context. Several of these runtimes, such as NodeJS, Python and Java, have been updated to support execution on as containers on ei ...
```
same as above, please update the knative link here

pritidesai commented:
thanks @mrutkows for this PR, have left some minor comments, overall it looks great, we might want to address the overall structure, moving samples under `examples` like @chmouel mentioned 🙏

mrutkows commented:
@pritidesai Also, just removed demo-commands.sh which was a legacy script to run just the java demo.  In the future, we should look to see about creating a java-only pipeline (and perhaps for each of the language branches).  It would be great if we could eventually have what Simon has been pushing for which are "pipeline tasks" so we could have a master pipleine that runs each of the language "pipeline branches" as tasks (with switch logic).

mrutkows commented:
@chmouel Hi Chris, seeing if you wished to re-review post-changes.  Thanks!

mrutkows commented:
@pritidesai Hi Priti, I think that all your comments have been addressed; would you be able to re-review?

pritidesai commented:
> @pritidesai Hi Priti, I think that all your comments have been addressed; would you be able to re-review?

Sorry @mrutkows for delay, will review it today 🙏

pritidesai commented on openwhisk/README.md:
```diff
@@ -1,262 +1,569 @@
 ---
 
-## Troubleshooting
+#### Java
```
NIT: s/####/###/ like other runtimes NodeJs and Python

pritidesai commented:
Thanks @mrutkows for updates, overall `YAML` content and organization looks good to me including directories Tekton resources:

<img width="469" alt="Screen Shot 2020-07-09 at 10 31 25 PM" src="https://user-images.githubusercontent.com/206285/87119941-2e5a9e00-c234-11ea-8140-5d38c7be05d6.png">

And subdirectories for each runtime:

<img width="295" alt="Screen Shot 2020-07-09 at 10 33 36 PM" src="https://user-images.githubusercontent.com/206285/87120043-606c0000-c234-11ea-9105-15234e8196f9.png">

@chmouel please take a look whenever you get a chance 🙏

afrittoli commented:
@mrutkows as you rebase, would you mind adding a body to the commit message? (https://github.com/tektoncd/community/blob/master/standards.md#commit-messages) Thanks!

afrittoli commented:
@vdemeester @sthaha can we discover pipelines and tasks in subfolders with hub? 
I think it's really useful to have an E2E example of how to use a task or set of tasks, by I wonder how we can surface this kind of setup in the hub.

sthaha commented:
> @vdemeester @sthaha can we discover pipelines and tasks in subfolders with hub?

Yes and no, we can parse all files in a repo but I think the ReOrg TEP addresses the issue of computable file paths from kind/name/version information i.e. given a resource `task` that has name `openwhisk` and version `0.1` where should a tool look for the resource definitions? 

In TEP we chose kind/resource/version/resource.yaml as the structure and it requires the current PR to be broken into 

```
  /task/openwhisk/0.1/openwhisk.yaml
  /pipeline/openwhisk/0.1/openswhisk.yaml
  /condition/openwhisk/0.1/openwhisk.yaml
```

So that the hub can look for `tasks` under `task` and pipeline under `/pipeline` 

cc: @mrutkows 

I have no (strong) opinion on this structure and is open to  changing this structure if it suits us better but let's do it as an amendment to the existing [TEP](https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md).

> I think it's really useful to have an E2E example of how to use a task or set of tasks, by I wonder how we can surface this kind of setup in the hub.

That is a great question and one that we should address especially because with support for pipelines in hub (doesn't exist now) we would have to display/inform users of dependent tasks, conditions etc ...

sthaha commented:
> Thanks @mrutkows for updates, overall `YAML` content and organization looks good to me including directories Tekton resources:

@mrutkows @pritidesai unfortunately this breaks the organisation proposed in the TEP :( which requires files to be organised as `{resource}/{name}/{version}/{name}.yaml` 

e.g. `/pipeline/build-openwhisk-app/0.1/build-openwhisk-app.yaml` instead of the current structure `task/openwhisk/0.1/pipeline/pipeline-to-build-openwhisk-app.yaml` or move all files except `/task/openwhisk/0.1/openwhisk.yaml` to `/task/openwhisk/0.1/samples` 

Additionally, the files are missing  mandatory metadata as stated in https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md#versioning-resources 

viz. 
```yaml
metadata:
   labels:
        app.kubernetes.io/version:
  annotations:
       tekton.dev/pipelines.minVersion: "0.12.1"

spec:
  description: >- 
      one line summary of the resource
  
      Para describing the resource 
```

You may also add additional metadata such as `tags` and `displayName`

mrutkows commented:
@sthaha Please know that this PR predated the referenced TEP, and I did not place them under the "task" folder structure, but rebased this Monday to adapt to a change done (https://github.com/tektoncd/catalog/pull/416) on behalf of all existing content.  I will look at how to conform to the TEP, but we are effectively a pipeline that (hopefully in the future) may be able to run in its entirety as a "pipeline task" (term being used) and fear we are falling under "uncharted ground" in the proposal that was recently implemented.

After reading the TEP, it looks like there is discretion under "pipeline" top-level structure, if I refactor to move our existing sub-directory structure under:

/pipelines/openwhisk/0.1/
  /conditions
  /images
  /pipelines
  /tasks
  /samples
    /pipelineruns/xxx

Would that meet the criteria in your mind?

I just want to have a v1beta1 level pipeline that the openwhisk community can use as soon as possible as they are stuck on alpha tasks/samples (which are quite dated).

sthaha commented:
@mrutkows , based on our conversation on [Slack](https://tektoncd.slack.com/archives/CK9JGKS20/p1594800224144000?thread_ts=1594738572.135500&cid=CK9JGKS20), like @afrittoli pointed out the TEP is in `"proposed"` state, so we can go ahead and merge the PR and then later rearrange the files to conform to the TEP. 

Please note that having the mandatory annotations and the `spec.description` will help others discover the task on hub - https://ui-tekton-hub.apps.openshift-web.p0s5.p1.openshiftapps.com/


You may want to address the comment on slack if you haven't already. 

> @mrutkows I wonder if it would make sense to keep the shell script, service account and other non-tekton resources under a dedicated folder, since those are not something that will be directly consumable via registry API?

mrutkows commented:
@sthaha @pritidesai I have assumed that with the recent hub going live that we should indeed adapt towards the referenced TEP.  Therefore, the latest commit moves the openwhisk folder (catalog entry) under `/pipeline` instead of where it had been moved (i.e., `/task`).  Additionally, I moved the pipeline resource YAML to the base directory and added the proper tags so that the Hub will have the correct description and search terms (along with min. versions).  

I look forward to being the first "pipeline" under the new structure and figuring out how we evolve towards the best practices for pipelines and their often many dependencies and considerations beyond that of a single task.

mrutkows commented:
@pritidesai @chmouel @sbwsg @sthaha I would truly appreciate a review in hope that after addressing all comments and rebasing/refactoring/adapting for the draft hub layout proposal that this finally can be merged.

pritidesai commented:
thanks @mrutkows for the changes

@sthaha please check @mrutkows reorganization, I have verified the content and all the YAML works. I have ran it and tested the pipeline.

sthaha commented:
@vdemeester @mrutkows Just noting that I am okay with merging the patch since the TEP is still in `proposed` but please be aware that the way it is currently organised doesn't [conform to proposal](https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md#organization) ([see example](https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md#example-of-a-resource-from-catalog))  and wouldn't show up in the Hub preview. 
This can be addressed in a separate patch of course.

I am also unsure about the `.tmpl` files. 

Since the pipeline is usable, I don't think we should bikeshed on it. However, we should discuss this further in our WG call. 

cc: @chmouel

chmouel commented:
/lgtm

thanks for filling #453 @sthaha  🙇 

Looking Good! 🤙🏽
